### ISSUE TITLE
Database schema suggestions

### ISSUE THREAD

**Issue by SlimeDog** [pmdevita/StayPut#5]
StayPut 1.2.x

Current schema:
```
CREATE TABLE `stayput_position` (
  `id` INTEGER PRIMARY KEY AUTOINCREMENT ,
  `uuid` VARCHAR ,
  `player_name` VARCHAR ,
  `world_name` VARCHAR ,
  `coordinate_x` DOUBLE PRECISION ,
  `coordinate_y` DOUBLE PRECISION ,
  `coordinate_z` DOUBLE PRECISION ,
  `yaw` FLOAT ,
  `pitch` FLOAT
  );
```
Suggested schema:
```
CREATE TABLE `stayput_position` (
  `uuid` VARCHAR(36) NOT NULL ,
  `world_name` VARCHAR( {suitably large integer}) NOT NULL ,
  `coordinate_x` REAL ,
  `coordinate_y` REAL ,
  `coordinate_z` REAL ,
  `yaw` REAL ,
  `pitch` REAL ,
  PRIMARY KEY ( uuid , world_name )
  );
```
Explanations:
- id attribute is unneeded, and allows substantively duplicate rows. The primary key should be ( uuid, world_name ). This would prevent duplicate rows (see insert/update statement below), without additional plugin logic.
- player_name attribute is unneeded. If a player name is ever needed, it should be a lookup on player UUID. The point of UUID is that the player name can change, but Mojang account identity does not.
- uuid attribute can never be null. It should be declared as
  - **VARCHAR(36) NOT NULL**
- world_name attribute can never be null. It should be declared as
  - **VARCHAR( {suitably large integer} ) NOT NULL**
- (Minor point) Floating point attributes should be declared as **REAL**, which is a synonym for DOUBLE PRECISION. Both SQLITE and MySQL accept this construction.

The database insert/update is the equivalent of
  - **INSERT OR REPLACE INTO stayput_position ( {columns} ) VALUES ( {values} ) ;**

**Comment by pmdevita** [comment]
Agreed, that's pretty much what I was thinking. I briefly looked into doing migrations earlier while I was working on fixing duplicates but ormlite doesn't have anything built in for that. Is there some way to give a table a version number or something so the plugin knows it's already done the changes or am I going to have to check against the schema?

**Comment by SlimeDog** [comment]
Check against the schema. There aren't that many of us yet. Do it soon. :)

**Comment by pmdevita** [comment]
Apparently ormlite does not support multiple primary keys https://github.com/ServiceStack/ServiceStack.OrmLite/#limitations

**Comment by edcous** [comment]
is this fixed?

**Comment by SlimeDog** [comment]
> is this fixed?

If you mean "has the schema changed?" then no, it had not changed as of StayPut 1.3.2-SNAPSHOT dev build, which was released on June 1.

**Comment by pmdevita** [comment]
It will be closed when it is fixed. This issue specifically will pretty much require an entire rewrite of the DB code to do it right so it will be some time before it is complete.

**Comment by pmdevita** [comment]
Fixed with 2.0.0 beta
